Machine ID: Add `bitbucket` join method for Bitbucket Pipelines joining

timothyb89 opened the pull request:
This adds a new `bitbucket` join method that Machine ID bots can use to authenticate to Teleport from Bitbucket Pipelines CI runs.

---

Minimal bitbucket pipeline config (using a custom build on my own infra with this patch):

```yaml
image: atlassian/default-image:3

pipelines:
  custom:
    run-tbot:
      - step:
          oidc: true
          script:
            - wget https://builds.cursed.industries/teleport-v17.0.0-dev-linux-amd64-bin.tar.gz
            - tar zxvf teleport-v17.0.0-dev-linux-amd64-bin.tar.gz
            - ./teleport/tbot start identity --storage=./tbot-data --destination ./tbot-user --join-method=bitbucket --proxy-server=example.cloud.gravitational.io:443 --token=bot-bitbucket --oneshot
            - ssh -F ./tbot-user/ssh_config root@b54e07fe397f.example.cloud.gravitational.io echo "hello world"
```

And associated `bitbucket`-type join token:
```yaml
kind: token
version: v2
metadata:
  # name will be specified in the `tbot` to use this token
  name: bot-bitbucket
spec:
  roles: [Bot]
  # bot_name should match the name of the bot created earlier in this guide.
  bot_name: bitbucket
  join_method: bitbucket
  bitbucket:
    identity_provider_url: 'https://api.bitbucket.org/2.0/workspaces/example/pipelines-config/identity/oidc'
    audience: 'ari:cloud:bitbucket::workspace/foo'
    allow:
    - workspace_uuid: '{foo}'
```

Fixes #28601

changelog: Support delegated joining for Bitbucket Pipelines in Machine ID

timothyb89 committed:
Add `bitbucket` join method for Bitbucket Pipelines joining

This adds a new `bitbucket` join method that Machine ID bots can use
to authenticate to Teleport from Bitbucket Pipelines CI runs.

zmb3 commented on lib/bitbucket/bitbucket.go:
```diff
@@ -0,0 +1,78 @@
+// audit events related to a specific join.
+func (c *IDTokenClaims) JoinAuditAttributes() (map[string]interface{}, error) {
+	res := map[string]interface{}{}
+	d, err := mapstructure.NewDecoder(&mapstructure.DecoderConfig{
```
Why not use encoding/json here?

strideynet replied:
Hm - from what I can recall, using map[string]interface{} allows us to put this into a protobuf audit event as structpb Struct, which is why I went for this. It feels a little odd to hold marshalled json inside of a protobuf.

tigrato replied:
the correct way of handling to structpb.Struct is to marshal it to json either using jsonpb or json in case it's gogo proto, and later unmarshal to the structpb.Struct

timothyb89 replied:
Yeah, looks like we do that properly when actually encoding into audit events or the database, using `apievents.EncodeMap()`: 
* `join.go`: https://github.com/gravitational/teleport/blob/master/lib/auth/join.go#L146
* `EncodeMap()`: https://github.com/gravitational/teleport/blob/master/api/types/events/struct.go#L101

I didn't fully understand the intent here before, but I think we're just trying to convert a struct to K/V pairs (i.e. `map[string]interface{}`) rather than to a JSON string, which is the internal representation we eventually pass along to `EncodeMap()`.

The struct was originally parsed from a JSON document (the JWT) so it already has tags to map to and from appropriate named fields; `mapstructure` is just a convenient way to do the conversion automatically instead of manually writing out the map.

I don't think `encoding/json` is able to help with that since we aren't going to/from a JSON string here.

strideynet replied:
> mapstructure is just a convenient way to do the conversion automatically instead of manually writing out the map.

Yeah - this maps to what I was thinking about when I originally added this. The goal was to be able to added metadata from the attestation process into the audit log. 90% of our join method are based on some kind of JWT, but we have a few which are not. 

I'm not particularly keen on the implementation I chose - but - if we're going to change this, we should address this in a separate PR/issue and update all of the join methods at once since this is in an interface on the join.go side.

zmb3 commented on lib/bitbucket/token_validator.go:
```diff
@@ -0,0 +1,89 @@
+
+// IDTokenValidator validates a Bitbucket issued ID Token.
+type IDTokenValidator struct {
+	IDTokenValidatorConfig
```
Should we be embedding here? Do we want to allow other packages to change the config fields after we've newed up the structure?

timothyb89 replied:
Hmm, fair. It's just vestigial, there's no reason `Clock` should be exposed at all. I'll inline it and make the field private.

timothyb89 committed:
Add unit tests for bitbucket joining

This also fully adds `deployment_environment_uuid` which was found to
be missing.

timothyb89 committed:
Fix imports

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/bitbucket-pipelines-joining

timothyb89 committed:
Update tf docs

timothyb89 committed:
Docs update

timothyb89 committed:
Update generated TF resources

strideynet commented on api/proto/teleport/legacy/types/types.proto:
```diff
@@ -1661,6 +1663,57 @@ message ProvisionTokenSpecV2TerraformCloud {
+    // targeted by this pipelines run, if any. These values may be found in the
+    // "Pipelines -> OpenID Connect -> Deployment environments" section of the
+    // repository settings.
+    string DeploymentEnvironmentUUID = 5 [(gogoproto.jsontag) = "deployment_environment,omitempty"];
```
Is this definitely the UUID? Their documentation gives some sample claims as :

```
{
  "sub": "{1de489be-ce6a-42a0-a8c8-eadbf1174ac7}:{bd715740-c970-486b-b68a-b421ec2a1f8b}:{759de0c6-eaee-4eaa-b7a6-c507eec759a7}",
  "aud": "ari:cloud:bitbucket::workspace/922e118c-f34f-4226-a897-0e0afbec553d",
  "stepUuid": "{759de0c6-eaee-4eaa-b7a6-c507eec759a7}",
  "deploymentEnvironment": "testing",
  "iss": "https://api.bitbucket.org/2.0/workspaces/bbcitest/pipelines-config/identity/oidc",
  "repositoryUuid": "{1de489be-ce6a-42a0-a8c8-eadbf1174ac7}",
  "branchName": "oidc-demo",
  "exp": 1578762300,
  "iat": 1578747600,
  "pipelineUuid": "{bee32f35-5dc0-4676-9ad1-657363be0ed0}",
  "workspaceUuid": "{922e118c-f34f-4226-a897-0e0afbec553d}"
}
```

If you've got a sample JWT to compare against, that'd be good.

timothyb89 replied:
Yeah, their documentation isn't quite right. They give an example payload in the OIDC page in the repo settings which _also_ isn't quite right, but it at least does show deployment env UUIDs. This is the `.well-known/openid-configuration`:

```json
{
  "authorization_endpoint": "urn:bitbucket_cloud:programmatic_authorization",
  "claims_supported": [
    "sub",
    "iss",
    "aud",
    "exp",
    "iat",
    "branchName",
    "compliant",
    "deploymentEnvironmentUuid",
    "pipelineUuid",
    "repositoryUuid",
    "stepUuid",
    "workspaceUuid"
  ],
  "id_token_signing_alg_values_supported": [
    "RS256"
  ],
  "issuer": "https://api.bitbucket.org/2.0/workspaces/while-mountain-mercury/pipelines-config/identity/oidc",
  "jwks_uri": "https://api.bitbucket.org/2.0/workspaces/while-mountain-mercury/pipelines-config/identity/oidc/keys.json",
  "response_types_supported": [
    "id_token"
  ],
  "subject_types_supported": [
    "public"
  ]
}
``` 

And some parsed token claims extracted from an actual run:
```json
{
  "aud": "ari:cloud:bitbucket::workspace/2c6775bc-e495-4ea3-a2b0-afb2b0a47a6c",
  "branchName": "master",
  "deploymentEnvironmentUuid": "{c2d993ea-9d0b-48f9-9a1e-53d35f7329b1}",
  "exp": 1731458335,
  "iat": 1731449335,
  "iss": "https://api.bitbucket.org/2.0/workspaces/while-mountain-mercury/pipelines-config/identity/oidc",
  "pipelineUuid": "{0b016cd1-7d58-4b93-b5d6-5c03e0a47931}",
  "repositoryUuid": "{727730ef-d2cd-4978-ae6b-be160266ac29}",
  "stepUuid": "{ae834543-bf7f-4468-9fe6-071f20ee874c}",
  "sub": "{727730ef-d2cd-4978-ae6b-be160266ac29}:{c2d993ea-9d0b-48f9-9a1e-53d35f7329b1}:{ae834543-bf7f-4468-9fe6-071f20ee874c}",
  "workspaceUuid": "{2c6775bc-e495-4ea3-a2b0-afb2b0a47a6c}"
}
```

strideynet replied:
Ok great - thanks TIm! A shame their docs are wrong...

strideynet commented on api/types/provisioning.go:
```diff
@@ -862,3 +876,31 @@ func (a *ProvisionTokenSpecV2TerraformCloud) checkAndSetDefaults() error {
 	return nil
 }
+
+func (a *ProvisionTokenSpecV2Bitbucket) checkAndSetDefaults() error {
```
Would be nice to include tests in provisioning_test.go for this.

timothyb89 replied:
Ah, good call, missed that file. Added!

strideynet reviewed (commented):
Generally looks good, just a few thoughts

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/bitbucket-pipelines-joining

timothyb89 committed:
Attempt to work around docs linter

timothyb89 committed:
Add provision token tests

timothyb89 commented on api/proto/teleport/legacy/types/types.proto:
```diff
@@ -1661,6 +1663,57 @@ message ProvisionTokenSpecV2TerraformCloud {
+    // repository settings.
+    string RepositoryUUID = 2 [(gogoproto.jsontag) = "repository_uuid,omitempty"];
+
+    // PipelineUUID is the UUID of the pipeline for which this token was issued.
+    // Bitbucket UUIDs must begin and end with braces, e.g. `{...}`
+    string PipelineUUID = 3 [(gogoproto.jsontag) = "pipeline_uuid,omitempty"];
+
+    // StepUUID is the UUID of the pipeline step for which this token was
+    // issued. Bitbucket UUIDs must begin and end with braces, e.g. `{...}`
+    string StepUUID = 4 [(gogoproto.jsontag) = "step_uuid,omitempty"];
```
I think I'm going to drop these 2 fields from the proto spec. They appear to be unique for every CI run so they're only useful as logged OIDC claims to identify the particular run/step.

timothyb89 committed:
Remove pipeline_uuid and step_uuid from protos

Also, fix deploment_environment_uuid field name.

timothyb89 committed:
Remove references to removed fields in tests

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/bitbucket-pipelines-joining

hugoShaka commented on lib/bitbucket/token_validator.go:
```diff
@@ -0,0 +1,82 @@
+func (id *IDTokenValidator) Validate(
+	ctx context.Context, issuerURL, audience, token string,
+) (*IDTokenClaims, error) {
+	p, err := oidc.NewProvider(ctx, issuerURL)
+	if err != nil {
+		return nil, trace.Wrap(err, "creating oidc provider")
+	}
```
Should we cache the provider for some time? Is there any risk of:
- DDoS-ing the bitbucket provider
- Getting rate-limited by the Bitbucket OIDC provider
- Exhausting our outbound connections (NAT saturationsx, too many go clients, ...)

How are the other join methods protecting against a join surge?

strideynet replied:
https://github.com/gravitational/teleport/issues/49213

hugoShaka reviewed (approved):
Approving because the JWKS caching issue is not specific to this PR. Every other join method that uses dynamically-fetched JWKS should cache its provider for a few minutes.

We might want to add a generic JWKS/oidc.Provider cache in a separate PR.

timothyb89 committed:
Switch to go-oidc/v3 and remove now-redundant nbf check

timothyb89 commented on lib/bitbucket/token_validator.go:
```diff
@@ -0,0 +1,82 @@
+	"context"
+	"time"
+
+	"github.com/coreos/go-oidc"
```
Looks like it was pretty much a drop in replacement for our use case, so I've pushed a change to switch over.

timothyb89 commented on lib/bitbucket/token_validator.go:
```diff
@@ -0,0 +1,82 @@
+		id.clock.Now(), time.Minute*2, idToken,
+	); err != nil {
+		return nil, trace.Wrap(err, "enforcing nbf")
+	}
```
Looks like it does. I've swapped to `go-oidc/v3` and removed the redundant check; unit tests continued to pass without changes which was nice to see.

timothyb89 committed:
Fix go.mod imports for the TF provider

timothyb89 committed:
Fix event-handler go.mod

timothyb89 committed:
Address review feedback; add 15s timeout to fetch provider metadata

timothyb89 commented on lib/bitbucket/token_source.go:
```diff
@@ -0,0 +1,46 @@
+	return tok, nil
+}
+
+func NewIDTokenSource(getEnv envGetter) *IDTokenSource {
```
Good call, I just inlined the definition.

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/bitbucket-pipelines-joining

timothyb89 committed:
Update lib/bitbucket/token_validator.go

Co-authored-by: rosstimothy <39066650+rosstimothy@users.noreply.github.com>

timothyb89 committed:
Fix build after constant rename

timothyb89 commented on lib/bitbucket/token_validator.go:
```diff
@@ -0,0 +1,71 @@
+func (id *IDTokenValidator) Validate(
+	ctx context.Context, issuerURL, audience, token string,
+) (*IDTokenClaims, error) {
+	p, err := oidc.NewProvider(ctx, issuerURL)
```
(And verified the timeout works on my test cluster by setting it to an unrealistically low value and receiving a context deadline exceeded error while joining.)
